src/test/java/com/amazon/ion/StructTest.java

```diff
@@ -1553,7 +1609,48 @@ public void testRandomChanges()
                 System.out.println("---");
             }
         }
+        // Now make sure the generated struct can be accessed concurrently by multiple threads after being made
+        // read-only.
+        testMultithreadedAccess(s1);
     }
+
+    /**
+     * Verifies that the givens struct can be accessed concurrently after it is made read-only.
+     * @param original the struct to test.
+     */
+    private void testMultithreadedAccess(IonStruct original) {
+        for (int i = 0; i < 2; i++) {
+            Map<String, List<IonValue>> results = new ConcurrentHashMap<>();
+            original.iterator().forEachRemaining(value -> results.put(value.getFieldName(), Collections.synchronizedList(new ArrayList<>())));
+            Set<String> fieldNames = results.keySet();
+            // One trial in which clone is performed first, the other in which it is not.
+            IonStruct struct = (i == 0) ? original : original.clone();
+            struct.makeReadOnly();
+            // Concurrently access all fields in the struct using many threads, recording the results in a concurrent
+            // multi-map.
```

Comment:
Will the guards in the `ConcurrentHashMap` and the synchronized list have any effects on the execution of the `CompletableFutures`?

If so, you might want to consider something like this, and have task `i` store its results directly into `results[i]`.
```java
Map<String, IonValue>[] results = new HashMap<String, IonValue>[numberOfTasks];
```